docs/configuration.rst
Configure the Bundle
====================

The bundle allows you to specify different Message Queue providers - however,
Amazon AWS and IronMQ are the only ones currently supported. Blocking, synchronous queues
are also supported through the ``sync`` driver to aid development and debugging.

We are actively looking to add more and would be more than happy to accept contributions.

Providers
---------

This bundle allows you to configure and use multiple supported providers with in the same
application. Each queue that you create is attached to one of your registered providers
and can have its own configuration options.

Providers may have their own dependencies that should be added to your ``composer.json`` file.

For specific instructions on how to configure each provider, please view their documents.

.. toctree::
    :maxdepth: 2

    aws-provider
    iron-mq-provider
    sync-provider

Caching
-------

Providers can leverage a caching layer to limit the amount of calls to the Message Queue
for basic lookup functionality - this is important for things like AWS's ARN values, etc.

By default the library will attempt to use file cache, however you can pass your
own cache service, as long as its an instance of ``Doctrine\Common\Cache\Cache``.

The configuration parameter ``cache_service`` expects the container service id of a registered
Cache service. See below.

.. code-block:: yaml

    #app/config.yml

    services:
        my_cache_service:
            class: My\Caching\CacheService

    uecode_qpush:
        cache_service: my_cache_service

**Note:** *Though the Queue Providers will attempt to create queues if they do not exist when publishing or receiving messages,
it is highly recommended that you run the included console command to build queues and warm cache from the CLI beforehand.*

Queue Options
-------------

Each queue can have their own options that determine how messages are published or received.
The options and their descriptions are listed below.

+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| Option                   | Description                                                                               | Default Value |
+==========================+===========================================================================================+===============+
| ``push_notifications``   | Whether or not to POST notifications to subscribers of a Queue                            | ``false``     |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``notification_retries`` | How many attempts notifications are resent in case of errors - if supported               | ``3``         |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``message_delay``        | Time in seconds before a published Message is available to be read in a Queue             | ``0``         |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``message_timeout``      | Time in seconds a worker has to delete a Message before it is available to other workers  | ``30``        |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``message_expiration``   | Time in seconds that Messages may remain in the Queue before being removed                | ``604800``    |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``messages_to_receive``  | Maximum amount of messages that can be received when polling the queue                    | ``1``         |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``receive_wait_time``    | If supported, time in seconds to leave the polling request open - for long polling        | ``3``         |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+
| ``subscribers``          | An array of Subscribers, containing an ``endpoint`` and ``protocol``                      | ``empty``     |
+--------------------------+-------------------------------------------------------------------------------------------+---------------+

Symfony Application as a Subscriber
-----------------------------------

The QPush Bundle uses a Request Listener which will capture and dispatch notifications from your queue providers for you. The specific route you use does not matter.

In most cases, it is recommended to just list the host or domain for your Symfony application as the ``endpoint`` of your subscriber.  You do not need to create a new action for QPush to receive messages.

Logging with Monolog
--------------------

By default, logging is enabled in the Qpush Bundle and uses Monolog, configured
via the MonologBundle. You can toggle the logging behavior by setting
``logging_enabled`` to ``false``.

Logs will output to your default Symfony environment logs using the 'qpush' channel.

Example Configuration
---------------------

A working configuration would look like the following

.. code-block:: yaml

    uecode_qpush:
        cache_service: null
        logging_enabled: true
        providers:
            aws_east_provider:
                driver: aws #optional for providers named 'aws' or 'ironmq'
                key: YOUR_AWS_KEY_HERE
                secret: YOUR_AWS_SECRET_HERE
                region: us-east-1
            aws_west_provider:
                driver: aws #required for named providers
                key: YOUR_AWS_KEY_HERE
                secret: YOUR_AWS_SECRET_HERE
                region: us-west-1
            ironmq_provider:
                driver: aws #optional for providers named 'aws' or 'ironmq'
                token: YOUR_IRONMQ_TOKEN_HERE
                project_id: YOUR_IRONMQ_PROJECT_ID_HERE
            dev_provider:
                driver: sync
        queues:
            my_queue_name:
                provider: ironmq_provider #or aws_east_provider etc
                options:
                    push_notifications:     true
                    notification_retries:   3
                    message_delay:          0
                    message_timeout:        30
                    message_expiration:     604800
                    messages_to_receive:    1
                    receive_wait_time:      3
                    subscribers:
                        - { endpoint: http://example1.com/, protocol: http }
                        - { endpoint: http://example2.com/, protocol: http }


src/DependencyInjection/Configuration.php
<?php

/**
 * Copyright 2014 Underground Elephant
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @package     qpush-bundle
 * @copyright   Underground Elephant 2014
 * @license     Apache License, Version 2.0
 */

namespace Uecode\Bundle\QPushBundle\DependencyInjection;

use Symfony\Component\Config\Definition\Builder\TreeBuilder;
use Symfony\Component\Config\Definition\ConfigurationInterface;

/**
 * @author Keith Kirk <kkirk@undergroundelephant.com>
 */
class Configuration implements ConfigurationInterface
{
    public function getConfigTreeBuilder()
    {
        $treeBuilder = new TreeBuilder();
        $rootNode    = $treeBuilder->root('uecode_qpush');

        $rootNode
            ->addDefaultsIfNotSet()
            ->children()
                ->scalarNode('cache_service')
                    ->defaultNull()
                ->end()
                ->booleanNode('logging_enabled')
                    ->defaultTrue()
                ->end()
                ->append($this->getProvidersNode())
                ->append($this->getQueuesNode())
            ->end()
        ;

        return $treeBuilder;
    }

    private function getProvidersNode()
    {
        $treeBuilder  = new TreeBuilder();
        $node         = $treeBuilder->root('providers');
        $requirements = [
            'aws' => ['key', 'secret'],
            'ironmq' => ['token', 'project_id'],
            'sync' => [],
        ];

        $node
            ->useAttributeAsKey('name')
            ->prototype('array')
                ->treatNullLike([])
                ->children()
                    ->enumNode('driver')
                        ->isRequired()
                        ->values(array_keys($requirements))
                    ->end()
                    // IronMQ
                    ->scalarNode('token')->end()
                    ->scalarNode('project_id')->end()
                    ->enumNode('host')
                        ->defaultValue('mq-aws-us-east-1')
                        ->values([
                            'mq-aws-us-east-1',
                            'mq-aws-eu-west-1',
                            'mq-rackspace-ord',
                            'mq-rackspace-lon',
                        ])
                    ->end()
                    ->scalarNode('port')
                        ->defaultValue('443')
                    ->end()
                    ->scalarNode('api_version')
                        ->defaultValue(1)
                    ->end()
                    // AWS
                    ->scalarNode('key')->end()
                    ->scalarNode('secret')->end()
                    ->scalarNode('region')
                        ->defaultValue('us-east-1')
                    ->end()
                ->end()

                ->validate()
                ->always()
                ->then(function (array $provider) use ($node, $requirements) {
                    foreach ($requirements[$provider['driver']] as $requirement) {
                        if (empty($provider[$requirement])) {
                            throw new \InvalidArgumentException(
                                sprintf('%s queue providers must have a %s; none provided', $provider['driver'], $requirement)
                            );
                        }
                    }

                    return $provider;
                })
            ->end()
        ;

        return $node;
    }

    private function getQueuesNode()
    {
        $treeBuilder = new TreeBuilder();
        $node        = $treeBuilder->root('queues');

        $node
            ->requiresAtLeastOneElement()
            ->useAttributeAsKey('name')
            ->prototype('array')
                ->children()
                    ->scalarNode('provider')
                        ->isRequired()
                        ->info('The Queue Provider to use')
                    ->end()
                    ->arrayNode('options')
                        ->children()
                            ->booleanNode('push_notifications')
                                ->defaultFalse()
                                ->info('Whether notifications are sent to the subscribers')
                            ->end()
                            ->scalarNode('notification_retries')
                                ->defaultValue(3)
                                ->info('How many attempts the Push Notifications are retried if the Subscriber returns an error')
                                ->example(3)
                            ->end()
                            ->scalarNode('notification_retry_delay')
                                ->defaultValue(60)
                                ->info('Delay between each Push Notification retry in seconds')
                                ->example(3)
                                ->end()
                            ->scalarNode('message_delay')
                                ->defaultValue(0)
                                ->info('How many seconds before messages are inititally visible in the Queue')
                                ->example(0)
                            ->end()
                            ->scalarNode('message_timeout')
                                ->defaultValue(30)
                                ->info('How many seconds the Queue hides a message while its being processed')
                                ->example(30)
                            ->end()
                            ->scalarNode('message_expiration')
                                ->defaultValue(604800)
                                ->info('How many seconds a message is kept in Queue, the default is 7 days (604800 seconds)')
                                ->example(604800)
                            ->end()
                            ->scalarNode('messages_to_receive')
                                ->defaultValue(1)
                                ->info('Max amount of messages to receive at once - an event will be fired for each individually')
                                ->example(1)
                            ->end()
                            ->scalarNode('receive_wait_time')
                                ->defaultValue(3)
                                ->info('How many seconds to Long Poll when requesting messages - if supported')
                                ->example(3)
                            ->end()
                            ->append($this->getSubscribersNode())
                        ->end()
                    ->end()
                ->end()
            ->end()
        ;

        return $node;
    }

    private function getSubscribersNode()
    {
        $treeBuilder = new TreeBuilder();
        $node        = $treeBuilder->root('subscribers');

        $node
            ->prototype('array')
                ->children()
                    ->scalarNode('endpoint')
                        ->info('The url or email address to notify')
                        ->example('http://foo.bar/qpush/')
                    ->end()
                    ->enumNode('protocol')
                        ->values(['email', 'http', 'https'])
                        ->info('The endpoint type')
                        ->example('http')
                    ->end()
                ->end()
            ->end()
        ;

        return $node;
    }
}
